File: teammanager/teams.ts
import { randomUUID } from "node:crypto";
import { api, APIError, Header, Query } from "encore.dev/api";
import { Prisma } from "@prisma/client";
import { prisma } from "./prisma";
import { requireSiteAdmin, requirePermission, resolveActor } from "../auth/rbac";
import { isSiteAdmin } from "../auth/permissions";
import { isValidSlug } from "../lib/slugs";
import { StructKeyspace, expireInSeconds } from "encore.dev/storage/cache";
import { cluster } from "../cache";
import { ADMINISTRATOR_ROLE, ADMINISTRATOR_ROLE_LIMIT } from "../lib/constants";

export const TEAM_WITH_MEMBERS_INCLUDE = {
  members: {
    include: { user: { select: { name: true, vrchatUsername: true } } },
    orderBy: { createdAt: "asc" as const },
  },
} satisfies Prisma.OrganizationInclude;

export const teamCache = new StructKeyspace<{ id: string }, Team>(cluster, {
  keyPattern: "team/:id",
  defaultExpiry: expireInSeconds(300), // 5 minutes
});

function slugify(name: string): string {
  return name
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/(^-|-$)/g, "");
}

// A team is modelled as a better-auth organization (issue #6). Beyond the core
// organization fields it carries a set of aggregate "container" statistics that
// higher-level ranking features can populate.
export interface TeamStats {
  rankingAverage: number | null;
  pointsAverage: number | null;
  seasonRank: number | null;
  averagePointsPerEvent: number | null;
}

export interface TeamMemberSummary {
  userId: string;
  name: string;
  role: string;
}

export interface Team {
  id: string;
  name: string;
  slug: string;
  logo: string | null;
  stats: TeamStats;
  administratorSlotsRemaining: number;
  members: TeamMemberSummary[];
}

export interface TeamListItem {
  id: string;
  name: string;
  slug: string;
  logo: string | null;
  administratorSlotsRemaining: number;
  memberCount: number;
}

interface GetTeamParams {
  id: string;
}

// Returns a team with its members and aggregate statistics.
export const getTeam = api(
  { expose: true, method: "GET", path: "/api/teams/:id" },
  async ({ id }: GetTeamParams): Promise<Team> => {
    const cached = await teamCache.get({ id });
    if (cached !== undefined) {
      return cached;
    }

    const organization = await prisma.organization.findUnique({
      where: { id },
      include: TEAM_WITH_MEMBERS_INCLUDE,
    });

    if (!organization) {
      throw APIError.notFound("team not found");
    }

    const team = toTeam(organization);
    await teamCache.set({ id }, team);
    return team;
  },
);

interface GetTeamBySlugParams {
  slug: string;
}

// Returns a team by its unique slug.
export const getTeamBySlug = api(
  { expose: true, method: "GET", path: "/api/teams/by-slug/:slug" },
  async ({ slug }: GetTeamBySlugParams): Promise<Team> => {
    const organization = await prisma.organization.findUnique({
      where: { slug },
      include: TEAM_WITH_MEMBERS_INCLUDE,
    });

    if (!organization) {
      throw APIError.notFound("team not found");
    }

    return toTeam(organization);
  },
);

type OrganizationWithMembers = {
  id: string;
  name: string;
  slug: string;
  logo: string | null;
  rankingAverage: number | null;
  pointsAverage: number | null;
  seasonRank: number | null;
  averagePointsPerEvent: number | null;
  members: { id: string; userId: string; role: string; user: { name: string; vrchatUsername: string | null } }[];
};

interface ListTeamsParams {
  search?: Query<string>;
  limit?: Query<number>;
  offset?: Query<number>;
}

interface ListTeamsResponse {
  teams: TeamListItem[];
  total: number;
}

// Lists all teams with search and pagination support.
export const listTeams = api(
  { expose: true, method: "GET", path: "/api/teams" },
  async (params: ListTeamsParams): Promise<ListTeamsResponse> => {
    const { search, limit, offset } = params || {};
    const where: any = {};
    if (search) {
      where.OR = [
        { name: { contains: search, mode: "insensitive" } },
        { slug: { contains: search, mode: "insensitive" } },
      ];
    }

    const total = await prisma.organization.count({ where });

    const organizations = await prisma.organization.findMany({
      where,
      take: limit ?? undefined,
      skip: offset ?? undefined,
      include: {
        members: {
          select: { role: true },
        },
      },
      orderBy: { name: "asc" },
    });

    const teams = organizations.map((org) => {
      const administratorCount = org.members.filter(
        (member) => member.role === ADMINISTRATOR_ROLE,
      ).length;

      return {
        id: org.id,
        name: org.name,
        slug: org.slug,
        logo: org.logo,
        administratorSlotsRemaining: Math.max(
          ADMINISTRATOR_ROLE_LIMIT - administratorCount,
          0,
        ),
        memberCount: org.members.length,
      };
    });

    return { teams, total };
  }
);

interface CreateTeamParams {
  authorization: Header<"Authorization">;
  name: string;
  logo?: string | null;
}

// Creates a new team (organization). Gated by site administrator.
export const createTeam = api(
  { expose: true, auth: true, method: "POST", path: "/api/teams" },
  async ({ authorization, name, logo }: CreateTeamParams): Promise<Team> => {
    await requireSiteAdmin(prisma, authorization);

    const slug = slugify(name);
    try {
      const organization = await prisma.organization.create({
        data: {
          id: randomUUID(),
          name,
          slug,
          logo: logo ?? null,
          updatedAt: new Date(),
        },
        include: TEAM_WITH_MEMBERS_INCLUDE,
      });

      return toTeam(organization);
    } catch (error) {
      if (error instanceof Prisma.PrismaClientKnownRequestError && error.code === "P2002") {
        throw APIError.alreadyExists("team with this slug already exists");
      }
      throw error;
    }
  }
);

export function toTeam(organization: OrganizationWithMembers): Team {
  const administratorCount = organization.members.filter(
    (member) => member.role === ADMINISTRATOR_ROLE,
  ).length;

  return {
    id: organization.id,
    name: organization.name,
    slug: organization.slug,
    logo: organization.logo,
    stats: {
      rankingAverage: organization.rankingAverage,
      pointsAverage: organization.pointsAverage,
      seasonRank: organization.seasonRank,
      averagePointsPerEvent: organization.averagePointsPerEvent,
    },
    administratorSlotsRemaining: Math.max(
      ADMINISTRATOR_ROLE_LIMIT - administratorCount,
      0,
    ),
    members: organization.members.map((member) => ({
      userId: member.userId,
      name: member.user.vrchatUsername ?? member.user.name,
      role: member.role,
    })),
  };
}

interface UpdateTeamParams {
  id: string;
  authorization: Header<"Authorization">;
  name?: string;
  slug?: string;
  logo?: string | null;
}

// Updates a team's core metadata (name, slug, logo).
// Accessible to team administrators (via organization update permission) or site administrators.
export const updateTeam = api(
  { expose: true, auth: true, method: "PATCH", path: "/api/teams/:id" },
  async ({
    id,
    authorization,
    name,
    slug,
    logo,
  }: UpdateTeamParams): Promise<Team> => {
    const actor = await resolveActor(prisma, authorization);
    const hasOrgUpdatePermission = async () => {
      try {
        await requirePermission(prisma, {
          authorization,
          organizationId: id,
          resource: "organization",
          action: "update",
        });
        return true;
      } catch {
        return false;
      }
    };

    if (!isSiteAdmin(actor.siteRole) && !(await hasOrgUpdatePermission())) {
      throw APIError.permissionDenied("cannot update team metadata");
    }

    const existing = await prisma.organization.findUnique({ where: { id } });
    if (!existing) {
      throw APIError.notFound("team not found");
    }

    let nextSlug = existing.slug;
    if (slug !== undefined && slug !== existing.slug) {
      if (!isValidSlug(slug)) {
        throw APIError.invalidArgument("invalid slug format or length");
      }
      const collision = await prisma.organization.findUnique({ where: { slug } });
      if (collision) {
        throw APIError.alreadyExists("team slug is already in use");
      }
      nextSlug = slug;
    }

    const organization = await prisma.organization.update({
      where: { id },
      data: {
        name: name ?? undefined,
        slug: nextSlug,
        logo: logo === undefined ? undefined : logo,
        updatedAt: new Date(),
      },
      include: TEAM_WITH_MEMBERS_INCLUDE,
    });

    await teamCache.delete({ id });

    return toTeam(organization);
  },
);

File: teammanager/teams.go
package teammanager

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
	"strings"
	"time"

	"encore.dev/beta/errs"
	"encore.app/auth"
	"encore.app/teammanager/sqlc"
)

// --- getTeam (mirrors ts-legacy/teammanager/teams.ts getTeam) ---

func getTeam(ctx context.Context, id string) (*Team, error) {
	if cached, err := teamCache.Get(ctx, teamCacheKey{ID: id}); err == nil {
		team := cached
		return &team, nil
	}
	team, err := loadTeam(ctx, id)
	if err != nil {
		return nil, err
	}
	_ = teamCache.Set(ctx, teamCacheKey{ID: id}, *team)
	return team, nil
}

// --- getTeamBySlug (mirrors getTeamBySlug) ---

func getTeamBySlug(ctx context.Context, slug string) (*Team, error) {
	row, err := q().GetOrgBySlug(ctx, slug)
	if errors.Is(err, sql.ErrNoRows) {
		return nil, &errs.Error{Code: errs.NotFound, Message: "team not found"}
	}
	if err != nil {
		return nil, err
	}
	members, err := loadMemberRows(ctx, row.ID)
	if err != nil {
		return nil, err
	}
	return toTeam(toOrgBySlug(row), members), nil
}

// --- listTeams (mirrors listTeams) ---

type ListTeamsResponse struct {
	Teams []TeamListItem `json:"teams"`
	Total int            `json:"total"`
}

func listTeams(ctx context.Context, search string, limit, offset int) (*ListTeamsResponse, error) {
	var total int64
	var stubs []sqlc.ListTeamRowsRow
	var err error
	if search == "" {
		if total, err = q().CountTeams(ctx); err != nil {
			return nil, err
		}
		stubs, err = q().ListTeamRows(ctx, sqlc.ListTeamRowsParams{
			Column1: int32(limit),
			Offset:  int32(offset),
		})
		if err != nil {
			return nil, err
		}
	} else {
		if total, err = q().CountTeamsBySearch(ctx, search); err != nil {
			return nil, err
		}
		bySearch, err := q().ListTeamRowsBySearch(ctx, sqlc.ListTeamRowsBySearchParams{
			Column1: search,
			Column2: int32(limit),
			Offset:  int32(offset),
		})
		if err != nil {
			return nil, err
		}
		stubs = make([]sqlc.ListTeamRowsRow, 0, len(bySearch))
		for _, s := range bySearch {
			stubs = append(stubs, sqlc.ListTeamRowsRow{
				ID: s.ID, Name: s.Name, Slug: s.Slug, Logo: s.Logo,
			})
		}
	}
	teams := make([]TeamListItem, 0, len(stubs))
	for _, s := range stubs {
		counts, err := q().CountMembersAndAdmins(ctx, sqlc.CountMembersAndAdminsParams{
			OrganizationId: s.ID,
			Role:           auth.AdministratorRole,
		})
		if err != nil {
			return nil, err
		}
		slots := auth.AdministratorRoleLimit - int(counts.Count_2)
		if slots < 0 {
			slots = 0
		}
		var logo *string
		if s.Logo.Valid {
			logo = &s.Logo.String
		}
		teams = append(teams, TeamListItem{
			ID: s.ID, Name: s.Name, Slug: s.Slug, Logo: logo,
			AdministratorSlotsRemaining: slots,
			MemberCount:                 int(counts.Count),
		})
	}
	return &ListTeamsResponse{Teams: teams, Total: int(total)}, nil
}

// --- createTeam (mirrors createTeam; site-admin gated) ---

func createTeam(ctx context.Context, authorization, name string, logo *string) (*Team, error) {
	if _, err := auth.RequireSiteAdmin(ctx, authorization); err != nil {
		return nil, err
	}
	slug := slugifyTeamName(name)
	if _, err := q().OrgIDBySlug(ctx, slug); err == nil {
		return nil, &errs.Error{Code: errs.AlreadyExists, Message: "team with this slug already exists"}
	} else if !errors.Is(err, sql.ErrNoRows) {
		return nil, err
	}
	var logoVal sql.NullString
	if logo != nil {
		logoVal = sql.NullString{String: *logo, Valid: true}
	}
	id, err := q().CreateOrg(ctx, sqlc.CreateOrgParams{
		Name:      name,
		Slug:      slug,
		Logo:      logoVal,
		UpdatedAt: sql.NullTime{Time: time.Now().UTC(), Valid: true},
	})
	if isUniqueViolation(err) {
		return nil, &errs.Error{Code: errs.AlreadyExists, Message: "team with this slug already exists"}
	}
	if err != nil {
		return nil, err
	}
	return loadTeam(ctx, id)
}

// --- updateTeam (mirrors updateTeam) ---

// UpdateTeamParams carries optional metadata updates. A nil pointer leaves
// the field unchanged; ClearLogo resets the logo to NULL.
type UpdateTeamParams struct {
	Name      *string
	Slug      *string
	Logo      *string
	ClearLogo bool
}

func updateTeam(ctx context.Context, authorization, id string, p *UpdateTeamParams) (*Team, error) {
	actor, err := auth.ResolveActor(ctx, authorization)
	if err != nil {
		return nil, err
	}
	if !auth.IsSiteAdmin(actor.SiteRole) {
		if _, _, err := auth.RequirePermission(ctx, authorization, id, "organization", "update"); err != nil {
			return nil, &errs.Error{Code: errs.PermissionDenied, Message: "cannot update team metadata"}
		}
	}
	existingSlug, err := q().OrgSlugByID(ctx, id)
	if errors.Is(err, sql.ErrNoRows) {
		return nil, &errs.Error{Code: errs.NotFound, Message: "team not found"}
	}
	if err != nil {
		return nil, err
	}
	nextSlug := existingSlug
	if p.Slug != nil && *p.Slug != existingSlug {
		if !auth.IsValidSlug(*p.Slug) {
			return nil, &errs.Error{Code: errs.InvalidArgument, Message: "invalid slug format or length"}
		}
		if _, err := q().OrgIDBySlug(ctx, *p.Slug); err == nil {
			return nil, &errs.Error{Code: errs.AlreadyExists, Message: "team slug is already in use"}
		} else if !errors.Is(err, sql.ErrNoRows) {
			return nil, err
		}
		nextSlug = *p.Slug
	}
	set := []string{`"slug" = $1`, `"updatedAt" = $2`}
	args := []any{nextSlug, time.Now().UTC()}
	if p.Name != nil {
		args = append(args, *p.Name)
		set = append(set, fmt.Sprintf(`"name" = $%d`, len(args)))
	}
	if p.ClearLogo {
		set = append(set, `"logo" = NULL`)
	} else if p.Logo != nil {
		args = append(args, *p.Logo)
		set = append(set, fmt.Sprintf(`"logo" = $%d`, len(args)))
	}
	args = append(args, id)
	_, err = db.Exec(ctx,
		`UPDATE "organization" SET `+strings.Join(set, ", ")+fmt.Sprintf(` WHERE id = $%d`, len(args)),
		args...,
	)
	if isUniqueViolation(err) {
		return nil, &errs.Error{Code: errs.AlreadyExists, Message: "team slug is already in use"}
	}
	if err != nil {
		return nil, err
	}
	invalidateTeamCache(ctx, id)
	return loadTeam(ctx, id)
}

// --- HTTP endpoints (thin wrappers over the cores above) ---

//encore:api public method=GET path=/api/teams/:id
func (s *Service) GetTeam(ctx context.Context, id string) (*Team, error) {
	return getTeam(ctx, id)
}

// GetTeamBySlugParams carries the slug query param. It lives at
// /api/teams-by-slug (rather than nested under /api/teams/) because static
// sub-paths route-conflict with /api/teams/:id under Encore's router.
type GetTeamBySlugParams struct {
	Slug string `query:"slug"`
}

//encore:api public method=GET path=/api/teams-by-slug
func (s *Service) GetTeamBySlug(ctx context.Context, p *GetTeamBySlugParams) (*Team, error) {
	return getTeamBySlug(ctx, p.Slug)
}

// ListTeamsRequest carries search/pagination query params.
type ListTeamsRequest struct {
	Search string `query:"search"`
	Limit  int    `query:"limit"`
	Offset int    `query:"offset"`
}

//encore:api public method=GET path=/api/teams
func (s *Service) ListTeams(ctx context.Context, p *ListTeamsRequest) (*ListTeamsResponse, error) {
	return listTeams(ctx, p.Search, p.Limit, p.Offset)
}

// CreateTeamRequest carries the auth header plus the new team's fields.
type CreateTeamRequest struct {
	Authorization string  `header:"Authorization"`
	Name          string  `json:"name"`
	Logo          *string `json:"logo,omitempty"`
}

//encore:api public method=POST path=/api/teams
func (s *Service) CreateTeam(ctx context.Context, p *CreateTeamRequest) (*Team, error) {
	return createTeam(ctx, p.Authorization, p.Name, p.Logo)
}

// UpdateTeamRequest carries the target id, auth header, and editable fields.
// The id travels in the body because this Encore version only accepts scalar
// params alongside path params.
type UpdateTeamRequest struct {
	ID            string  `json:"id"`
	Authorization string  `header:"Authorization"`
	Name          *string `json:"name,omitempty"`
	Slug          *string `json:"slug,omitempty"`
	Logo          *string `json:"logo,omitempty"`
	ClearLogo     bool    `json:"clearLogo,omitempty"`
}

//encore:api public method=PATCH path=/api/teams
func (s *Service) UpdateTeam(ctx context.Context, p *UpdateTeamRequest) (*Team, error) {
	return updateTeam(ctx, p.Authorization, p.ID, &UpdateTeamParams{
		Name: p.Name, Slug: p.Slug, Logo: p.Logo, ClearLogo: p.ClearLogo,
	})
}
